lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,260 @@
+      argNames.insert(argNames.begin(), args.begin(), args.end());
+
+    SmallVector<Value> newOperands;
+    newOperands.reserve(inputs.size());
+    SmallVector<std::tuple<StringAttr, Type, Location>> newInputs;
+    newInputs.reserve(inputs.size());
+    for (const auto &[name, value] : inputs) {
+      // If no argNames attribute, donot add it.
+      if (!argNames.empty())
+        argNames.push_back(name);
+      // Additional operands to the Instance op.
+      newOperands.push_back(value);
+      // Prepare the additional input ports that need to be added to the
+      // referenced module.
+      newInputs.emplace_back(name, value.getType(), value.getLoc());
+    }
+    // Add the new operands to the instance.
+    instance->insertOperands(instance->getNumOperands(), newOperands);
+    if (!argNames.empty())
+      instance->setAttr("argNames",
+                        ArrayAttr::get(instance->getContext(), argNames));
+    // Update all the modules with the new input ports.
+    for (auto refMod : refModules) {
+      refMod.appendPorts(newInputs, {});
+    }
+  }
+
+  /// Get the referenced modules for an instance.
+  SmallVector<HWModuleOp, 1> getReferencedModules(HWInstanceLike inst) {
+    SmallVector<HWModuleOp, 1> refMods;
+    for (auto refModName : inst.getReferencedModuleNamesAttr()) {
+      auto *node = graph->lookup(cast<StringAttr>(refModName));
+      if (auto instModule = dyn_cast_or_null<HWModuleOp>(*node->getModule()))
+        refMods.push_back(instModule);
+    }
+    return refMods;
+  }
+
+  /// If the `argNames` attribute exists, return it.
+  SmallVector<StringAttr> getInstanceArgNames(HWInstanceLike inst) {
+    SmallVector<StringAttr> argNames;
+    if (auto args = inst->getAttrOfType<ArrayAttr>("argNames")) {
+      auto attr = args.getAsRange<StringAttr>();
+      argNames.insert(argNames.begin(), attr.begin(), attr.end());
+    }
+    return argNames;
+  }
+
+  /// Combine two consecutive clock_gate ops, into one.
+  void collapseConsecutiveClockGates(ClockGateOp driverGate,
+                                     ClockGateOp userGate) {
+    mlir::OpBuilder builder(context);
+    builder.setInsertionPoint(userGate);
+    Value enable = getEnable(driverGate);
+    enable = builder.create<comb::AndOp>(userGate->getLoc(), enable,
+                                         userGate.getEnable());
+    userGate.getInputMutable().assign(driverGate.getInput());
+    userGate.getEnableMutable().assign(enable);
+  }
+
+  /// If the clock_gate op has `test_enable`, combine it with the `enable` and
+  /// return the final enable.
+  Value getEnable(ClockGateOp clkGate) {
+    Value enable = clkGate.getEnable();
+    // If there is testEnable, or it.
+    if (auto testEnable = clkGate.getTestEnable()) {
+      OpBuilder builder(clkGate.getContext());
+      builder.setInsertionPointAfter(clkGate);
+      enable =
+          builder.create<comb::OrOp>(clkGate->getLoc(), enable, testEnable);
+    }
+    return enable;
+  }
```

Comment:
I'm wondering if we'd need to keep the regular `enable` and the `test_enable` separate… probably not?

Reply:
I asked David, and seems like its okay to combine them. Not sure of other use cases, if that's required then we need two additional ports for every gated clock.

Reply:
Ah right, good point, forgot about that. I was just wondering of how a user might expect the pass to behave in CIRCT -- someone could expect that the enable inputs would be preserved as separate signals. But on the other hand this is also very easy to adjust once the pattern of the pass is established, and this comes up as a concrete need. Combining them feels like a good start!